### ISSUE TITLE
mail: add send_after_commit option

### ISSUE THREAD

**Issue by yelizariev** [odoo/odoo#59585]
This is a potential TODO for master.

We normally don't want mails be sent if transaction is aborted. Currently, you need to wrap a function and pass it to ``cr.after`` each time you want to postpone mails until the end of transaction.

Also, sending emails may force making commit immediately, which may break workflow if transaction is aborted after sending email. See https://github.com/odoo/odoo/pull/58804 for details.

Something similar is done for notifications, but it works only if there is more than 50 followers:
https://github.com/odoo/odoo/blob/341639521ff916afa23abd2b10bbb1e9ce72ed58/addons/mail/models/mail_thread.py#L2326-L2331

It should be available in other methods like ``message_post_with_template``, ``message_post``.

**Comment by yelizariev** [comment]
More general method: https://github.com/odoo/enterprise/blob/afa8412c937c20ceabadfb72ffc4cacc22aaf528/l10n_mx_edi/tools/run_after_commit.py#L12

**Comment by yelizariev** [comment]
@rco-odoo 
I think to add [run_after_commit](https://github.com/odoo/enterprise/blame/afa8412c937c20ceabadfb72ffc4cacc22aaf528/l10n_mx_edi/tools/run_after_commit.py#L12) decorator to Odoo 13. It's needed to postpone email/sms: https://github.com/odoo/odoo/pull/59965#issuecomment-709319665

Do you think it's ok to add it in stable branch? Where is a proper place for it? 
* odoo/api.py
* odoo/tools/....

With that decorator, PR https://github.com/odoo/odoo/pull/59965 will be just applying the decorator for this method:

https://github.com/odoo/odoo/blob/063e617a832d582d460dee0fcb54d837f3cafb1c/addons/stock/models/stock_picking.py#L776-L779

**Comment by yelizariev** [comment]
Also, will this decorator work when we have method inheritance (like here https://github.com/odoo/odoo/pull/59965#issuecomment-709319665)

**Comment by xmo-odoo** [comment]
> Do you think it's ok to add it in stable branch? 

Not unless it's absolutely necessary.

> Also, will this decorator work when we have method inheritance (like here [#59965 (comment)](https://github.com/odoo/odoo/pull/59965#issuecomment-709319665))?

Not even remotely.

**Comment by xmo-odoo** [comment]
> We normally don't want mails be sent if transaction is aborted. Currently, you need to wrap a function and pass it to `cr.after` each time you want to postpone mails until the end of transaction.

Looking at the code it looks like `message_post_with_template` just... goes into mass_mail mode which triggers an immediate mode, that seems pretty sus, maybe just passing in `composition_mode='comment'` would suffice? /cc @tde-banana-odoo

**Comment by rco-odoo** [comment]
@yelizariev I think it is a bad idea to put such "async" decorators on business methods, because it prevents you from extending this method by overriding.  Also, the proposed decorator looks a bit *overkill* to me, although I am not 100% sure about this.

**Comment by tde-banana-odoo** [comment]
> > We normally don't want mails be sent if transaction is aborted. Currently, you need to wrap a function and pass it to `cr.after` each time you want to postpone mails until the end of transaction.
> 
> Looking at the code it looks like `message_post_with_template` can go into mass_mail mode which triggers an immediate send if more than one picking is involved? But there should not be more than one picking per send here so the `force_send` seems pretty sus /cc @tde-banana-odoo

I don't have the whole context; but yes, you don't want to use force_send, unless you really know what you do and are sure you don't send 1000 emails in your transaction. Normally just preparing mail.mail records and waiting for the cron to send them is already complex enough.

**Comment by yelizariev** [comment]
``force_send`` doesn't help with the problem of calling ``cr.commit()`` on generating pdf. See #59965

**Comment by nim-odoo** [comment]
If I'm not misleading, our use case is a bit more subtle:
- The default mail template is `mail_template_data_delivery_confirmation`, which has an attachment:
https://github.com/odoo/odoo/blob/294628df479fba4db5e7859f1fc2d2b13bae8c6f/addons/stock/data/mail_template_data.xml#L39
- When generating a PDF, we commit:
https://github.com/odoo/odoo/blob/0f885721d9902140c9feca4025e2b35698f9d1b9/odoo/addons/base/models/ir_actions_report.py#L675-L685
https://github.com/odoo/odoo/blob/0f885721d9902140c9feca4025e2b35698f9d1b9/odoo/addons/base/models/assetsbundle.py#L304-L305

So the commit it triggered by the fact that we have an attachment, not the fact that we send the mail.

To be checked: do we really need to commit for our specific use case, and more generally when calling `message_post_with_template`?

**Comment by sebalix** [comment]
@nim-odoo indeed, and I was asking here if we could do that process in a separate transaction?
https://github.com/odoo/odoo/pull/58804#issuecomment-700672281

**Comment by odony** [comment]
> To be checked: do we really need to commit for our specific use case, and more generally when calling `message_post_with_template`?

Yes we do, unfortunately. And being in the process of sending an email doesn't make much difference. A commit is always disruptive in the middle of a transaction, because it leaves things in an inconsistent state.

There are hardly any other option in this case (printing a PDF report + obsolete assets bundles) because:
- wkhtmltopdf will load assets in its own transaction, so the attachments need to be committed first, or the PDF will be rendered without assets (the assets URLs are rendered in the main transaction)
- the commit happens *if and only if* the bundle attachments were outdated, which a rare one-off occurrence, but it does happen
- you can't open a different cursor in `save_attachment()` to commit the bundles without committing the main transaction because you have a chance to make the problem worse:
  + Nice case: `save_attachment()` fails with a concurrent update error, you let it bubble up, and the whole request is retried. In practice you can't fully control that, as some override could swallow the exception.
  + Bad case: instead of a concurrent update error, `save_attachment()` blocks due to an interleaved lock with the main transaction. PG can't detect and break this deadlock, so it will have to timeout. 

And yes, the transaction problems with additional cursors are *extremely likely* in practice on busy systems, because when asset invalidation occurs many workers will be fighting to re-generate the new bundle.

---

Taking a step back, it seems the problem being discussed is a very rare corner case: a spurious commit that can *sometimes* break transaction integrity if all of the following happens *at the same time*:
1) a PDF report is rendered as part of another business transaction
2) the report bundles were recently invalidated by a source code update
3) that commit causes an actual integrity problem because the business transaction fails *after* rendering the report (which means that the transaction continues for significant operations after printing the PDF, and fails)

I'm not saying there is no solution to this problem, but the chances to introduce bigger problems than the original issue are very high. We'd be touching core mechanisms to fix a very minor problem. That's never a good outlook for a stable patch.

**Comment by yelizariev** [comment]
Commiting can be avoided, if wkhtmltopdf doesn't need sending requests to Odoo to load resources (css). I.e. in main transaction, we prepare html and assets, save all files on disk, run wkhtmltopdf, and then load pdf _into_ main transaction.
 But indeed, it's only for master and only to fix situation when all 3 conditions met
